internal/cadence/linter.go

```diff
@@ -0,0 +1,351 @@
+func (l *linter) decideCheckerConfig(program *ast.Program) *sema.Config {
+	if program.SoleTransactionDeclaration() != nil || program.SoleContractDeclaration() != nil {
+		return l.checkerStandardConfig
+	}
+
+	return l.checkerScriptConfig
+}
+
+func (l *linter) handleImport(
+	checker *sema.Checker,
+	importedLocation common.Location,
+	_ ast.Range,
+) (
+	sema.Import,
+	error,
+) {
+	switch importedLocation {
+	case stdlib.CryptoCheckerLocation:
+		cryptoChecker := stdlib.CryptoChecker()
+		return sema.ElaborationImport{
+			Elaboration: cryptoChecker.Elaboration,
+		}, nil
+	case stdlib.TestContractLocation:
+		testChecker := stdlib.GetTestContractType().Checker
+		return sema.ElaborationImport{
+			Elaboration: testChecker.Elaboration,
+		}, nil
+	case cdcTests.BlockchainHelpersLocation:
+		helpersChecker := cdcTests.BlockchainHelpersChecker()
+		return sema.ElaborationImport{
+			Elaboration: helpersChecker.Elaboration,
+		}, nil
+	default:
+		filepath, err := l.resolveImportFilepath(importedLocation, checker.Location)
+		if err != nil {
+			return nil, err
+		}
+
+		importedChecker, ok := l.checkers[filepath]
+		if !ok {
+			code, err := l.state.ReadFile(filepath)
+			if err != nil {
+				return nil, err
+			}
+
+			importedProgram, err := parser.ParseProgram(nil, code, parser.Config{})
+
+			if err != nil {
+				return nil, err
+			}
+			if importedProgram == nil {
+				return nil, &sema.CheckerError{
+					Errors: []error{fmt.Errorf("cannot import %s", importedLocation)},
+				}
+			}
+
+			importedChecker, err = checker.SubChecker(importedProgram, importedLocation)
+			if err != nil {
+				return nil, err
+			}
+
+			l.checkers[filepath] = importedChecker
+			err = importedChecker.Check()
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		return sema.ElaborationImport{
+			Elaboration: importedChecker.Elaboration,
+		}, nil
+	}
+}
+
+func (l *linter) resolveImportFilepath(
+	location common.Location,
+	parentLocation common.Location,
+) (
+	string,
+	error,
+) {
+	switch location := location.(type) {
+	case common.StringLocation:
+		// If the location is not a cadence file try getting the code by identifier
+		if !strings.Contains(location.String(), ".cdc") {
+			contract, err := l.state.Contracts().ByName(location.String())
+			if err != nil {
+				return "", err
+			}
+
+			return contract.Location, nil
+		}
+
+		// If the location is a cadence file, resolve relative to the parent location
+		parentPath := ""
+		if parentLocation != nil {
+			parentPath = parentLocation.String()
+		}
+
+		resolvedPath := filepath.Join(filepath.Dir(parentPath), location.String())
+		return resolvedPath, nil
+	default:
+		return "", fmt.Errorf("unsupported location: %T", location)
+	}
+}
+
+// helpers
+
+func getDiagnosticsForParsingCheckingError(
+	err error,
+	location common.Location,
+	code string,
+) (
+	[]analysis.Diagnostic,
+	error,
+) {
+	parentErr, ok := err.(errors.ParentError)
```

Comment:
lets use `errors.As` if we can as that is preferred. https://pkg.go.dev/errors#section-documentation
```
var perr *fs.PathError
if errors.As(err, &perr) {
	fmt.Println(perr.Path)
}
is preferable to

if perr, ok := err.(*fs.PathError); ok {
	fmt.Println(perr.Path)
}
because the former will succeed if err wraps an [*io/fs.PathError](https://pkg.go.dev/io/fs#PathError).
```